feat: add CacheStore for GitHub stats and use names instead of usernames

CMEONE opened the pull request:
This pull request makes the following changes:
- moves the `CachedData` type to its own file
- creates `CacheStore` to store generalized cached data (currently only used for `github` contributor data)
- refactors `GitHubStatsService` to use `CacheStore` instead of previous nonpersistent `cache` instance variable
- fetches and caches the names of contributors (if a name is associated with the account)
- uses names (where available) instead of usernames in the contributor page to increase consistency with hard-coded contributors

According to [GitHub documentation](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api):
> The primary rate limit for unauthenticated requests is 60 requests per hour

As long as the number of contributors does not exceed 58-59 (depending on `includeMergedPRs`), there won't be any rate limit issues as the cache is only updated after an hour of staleness. Requests for contributor stats (and merged PRs if `includeMergedPRs`) are prioritized before fetching contributor names, so if a rate limit gets hit while fetching contributor names, the usernames for the remaining contributors are used as a fallback, and the `GitHubStatsService` will continue to attempt fetching the remaining contributors each time stats are requested again (when the settings & credits page is loaded/reloaded).

Eventually, it might make more sense to cache GitHub-related data in the repository itself (through GitHub actions), but this solution is viable at the current number of contributors and gracefully handles situations where rate limits occur.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/405)
<!-- Reviewable:end -->

doprz commented on src/shared/storage/CacheStore.ts:
```diff
@@ -0,0 +1,18 @@
+
+interface ICacheStore {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    github: Record<string, CachedData<any>>;
```
Use `unknown` instead of `any` here and narrow down the type later on.

doprz commented on src/views/lib/getGitHubStats.ts:
```diff
@@ -68,24 +66,30 @@ export type LD_ADMIN_GITHUB_USERNAMES = (typeof LONGHORN_DEVELOPERS_ADMINS)[numb
-        if (cachedItem && Date.now() - cachedItem.dataFetched.getTime() < CACHE_TTL) {
+    private async getCachedData<T>(key: string): Promise<CachedData<T> | null> {
+        if(Object.keys(this.cache).length === 0) {
+            this.cache = await CacheStore.get('github') as Record<string, CachedData<any>>;
```
Same here

doprz commented on src/views/lib/getGitHubStats.ts:
```diff
@@ -68,24 +66,30 @@ export type LD_ADMIN_GITHUB_USERNAMES = (typeof LONGHORN_DEVELOPERS_ADMINS)[numb
+            this.cache = await CacheStore.get('github') as Record<string, CachedData<any>>;
+        }
+        const cachedItem = this.cache[key];
+        if (cachedItem && Date.now() - new Date(cachedItem.dataFetched).getTime() < CACHE_TTL) {
```
Refactor this to be cleaner such as by storing the date difference in a const.

doprz reviewed (changes requested):
Overall great PR that adheres to our code standards and is well documented. I've requested a couple of changes in addition that all checks must pass.

DereC4 commented:
any updates?

CMEONE commented:
Due to concerns over rate limiting issues, I have refactored the fetch pipeline to utilize a cached version of the GitHub API as the primary API call (with a fallback to the regular GitHub API in the rare chance that the cached endpoint is not available) through the `fetchGitHub` method. I have updated `fetchContributorStats` and `fetchContributorNames` to use `fetchGitHub` internally (which pretty much entirely avoids GitHub rate limits due to the cached API endpoint).

I believe this solution works well for now as it solves the issue of rate limiting while simultaneously providing live-updating contribution stats (without needing a new version of the extension to include updated stats on build). It is worth noting that the current version of the extension (before this PR) calls the unauthenticated GitHub API to fetch contributor stats which would cause a rate limit issue after 60 page loads per IP address, so even if a different solution gets adopted in the future, this one can mitigate the short-term issue of rate limits (which exists in the current version of the extension).

doprz commented on src/views/lib/getGitHubStats.ts:
```diff
@@ -101,24 +119,31 @@ export class GitHubStatsService {
-                repo: REPO_NAME,
-            })
+        const data = await this.fetchWithRetry(() =>
+            this.fetchGitHub(`/repos/${REPO_OWNER}/${REPO_NAME}/stats/contributors`)
```
nit. Since this string template won't change, let's update it to a const.

CMEONE replied:
Do you recommend adding the constant to the constants section at the top of the file (right below `REPO_OWNER` and `REPO_NAME`) or right before the `fetchWithRetry` statement?

doprz replied:
I would add it to the constants section.

doprz commented on src/views/lib/getGitHubStats.ts:
```diff
@@ -101,24 +119,31 @@ export class GitHubStatsService {
         }
     }
 
+    private async fetchGitHub(route: string): Promise<unknown> {
+        try {
+            const response = await fetch(`https://github.cachedapi.com${route}`);
+            return await response.json();
+        } catch (error: unknown) {
+            const response = await fetch(`https://api.github.com${route}`);
+            return await response.json();
+        }
+    }
+
```
Use the URL Web API instead of ES6 string templates:
https://developer.mozilla.org/en-US/docs/Web/API/URL/URL

doprz reviewed (approved):
LGTM, thank you for writing clean code that adheres to our coding standards and for documenting this well! Approved!
